{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "ir",
      "display_name": "R"
    },
    "language_info": {
      "name": "R"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/jyryu3161/biobigdata/blob/main/transcriptome_analysis.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### <b>edgeR package를 이용한 transcriptome 분석</b>\n",
        "* normalized read count 사용"
      ],
      "metadata": {
        "id": "cPAEicJ1A_pM"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# 패키지 설치\n",
        "if (!require(\"BiocManager\", quietly = TRUE))\n",
        "    install.packages(\"BiocManager\")\n",
        "\n",
        "BiocManager::install(\"clusterProfiler\")\n",
        "BiocManager::install(\"enrichplot\")\n",
        "BiocManager::install(\"gprofiler2\")\n",
        "BiocManager::install(\"limma\")\n",
        "BiocManager::install(\"edgeR\")\n",
        "install.packages(\"pheatmap\")\n",
        "install.packages(\"ggplot2\")\n",
        "install.packages(\"ggrepel\")"
      ],
      "metadata": {
        "id": "LyKF9FgA1YXn",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "7e37bec5-e5cc-4565-e820-cd71c6723ec7"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "\n",
            "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
            "'help(\"repositories\", package = \"BiocManager\")' for details.\n",
            "Replacement repositories:\n",
            "    CRAN: https://cran.rstudio.com\n",
            "\n",
            "Bioconductor version 3.19 (BiocManager 1.30.23), R 4.4.0 (2024-04-24)\n",
            "\n",
            "Installing package(s) 'BiocVersion', 'clusterProfiler'\n",
            "\n",
            "also installing the dependencies ‘zlibbioc’, ‘UCSC.utils’, ‘GenomeInfoDbData’, ‘formatR’, ‘XVector’, ‘GenomeInfoDb’, ‘lambda.r’, ‘futile.options’, ‘gridGraphics’, ‘tweenr’, ‘polyclip’, ‘RcppEigen’, ‘gridExtra’, ‘RcppArmadillo’, ‘lazyeval’, ‘plogr’, ‘png’, ‘Biostrings’, ‘futile.logger’, ‘snow’, ‘BH’, ‘cowplot’, ‘fastmatch’, ‘ggplotify’, ‘patchwork’, ‘ggplot2’, ‘ggforce’, ‘ggrepel’, ‘viridis’, ‘tidygraph’, ‘graphlayouts’, ‘ape’, ‘tidytree’, ‘treeio’, ‘BiocGenerics’, ‘Biobase’, ‘IRanges’, ‘RSQLite’, ‘S4Vectors’, ‘KEGGREST’, ‘HDO.db’, ‘BiocParallel’, ‘fgsea’, ‘reshape2’, ‘aplot’, ‘ggfun’, ‘ggnewscale’, ‘ggraph’, ‘scatterpie’, ‘shadowtext’, ‘ggtree’, ‘Rcpp’, ‘AnnotationDbi’, ‘downloader’, ‘DOSE’, ‘enrichplot’, ‘GO.db’, ‘GOSemSim’, ‘gson’, ‘igraph’, ‘plyr’, ‘qvalue’, ‘yulab.utils’\n",
            "\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# 1. 필요한 라이브러리를 불러옴\n",
        "library(edgeR)\n",
        "library(limma)\n",
        "library(clusterProfiler)\n",
        "library(enrichplot)\n",
        "library(ggplot2)\n",
        "library(gprofiler2)\n",
        "library(pheatmap)\n",
        "library(ggrepel)\n",
        "\n",
        "# 2. 데이터를 불러옴\n",
        "expr_data <- read.csv(\"transcriptome.csv\", row.names=1)\n",
        "expr_data <- expr_data[apply(expr_data, 1, var) != 0, ]\n",
        "design <- read.csv(\"study_design.csv\")\n",
        "\n",
        "# 3. DEG 분석을 수행\n",
        "group <- factor(design$condition) # study design 파일에서 group 정보를 불러옴\n",
        "y <- DGEList(counts=expr_data, group=group)\n",
        "y <- cpm(y)\n",
        "\n",
        "# t-test\n",
        "design_matrix <- model.matrix(~0+group) # design matrix 생성\n",
        "fit <- lmFit(y, design_matrix)\n",
        "contrast_matrix <- makeContrasts(Diff=grouptest-groupcontrol, levels=design_matrix) # group 이름 확인\n",
        "fit2 <- contrasts.fit(fit, contrast_matrix)\n",
        "fit2 <- eBayes(fit2)\n",
        "ttest_res <- topTable(fit2, number=Inf)\n",
        "\n",
        "# Multiple testing correction using FDR\n",
        "ttest_res$adj.P.Val <- p.adjust(ttest_res$P.Value, method=\"fdr\")\n",
        "\n",
        "# 5. 결과를 시각화\n",
        "# Volcano plot\n",
        "deg_results <- data.frame(log2FoldChange = ttest_res$logFC, p_adj = ttest_res$adj.P.Val, Gene_name = rownames(res))\n",
        "deg_results <- na.omit(deg_results)\n",
        "Category <- c()\n",
        "for(i in 1:length(deg_results$log2FoldChange)){\n",
        "  if((deg_results$log2FoldChange[i] < -1) & (deg_results$p_adj[i] < 0.05)){\n",
        "    Category[i] <- \"Down-regulated\"\n",
        "  }else if((deg_results$log2FoldChange[i] > 1) & (deg_results$p_adj[i] < 0.05)){\n",
        "    Category[i] <- \"Up-regulated\"\n",
        "  }else{\n",
        "    Category[i] <- \"None\"\n",
        "  }\n",
        "}\n",
        "deg_results[\"Category\"] <- Category\n",
        "deg_results$Category <- factor(deg_results$Category,\n",
        "                               levels = c(\"Down-regulated\", \"Up-regulated\", \"None\"),\n",
        "                               labels = c(\"blue\", \"red\", \"grey\"))\n",
        "\n",
        "top10_genes <- head(deg_results[order(deg_results$p_adj),], 10)\n",
        "top10_gene_names <- top10_genes$Gene_name\n",
        "\n",
        "my_plot_1 <- ggplot(data = deg_results, mapping = aes(x = log2FoldChange, y = -log10(p_adj))) +\n",
        "  geom_point(aes(colour = Category), size = 2) +\n",
        "  scale_color_identity(guide = 'legend', breaks = c(\"blue\", \"red\"), labels = c(\"Down-regulated\", \"Up-regulated\")) +\n",
        "  geom_hline(yintercept = -log10(0.05), linetype = \"dashed\") +\n",
        "  geom_vline(xintercept = c(-1, 1), linetype = \"dashed\") +\n",
        "  labs(x = \"log2 Fold Change\", y = \"-log10(p.adj)\", colour = NULL) +\n",
        "  ggtitle(\"Volcano Plot\") +\n",
        "  theme_minimal() +\n",
        "  theme(plot.title = element_text(hjust = 0.5)) +\n",
        "  geom_text_repel(data = top10_genes, aes(label = top10_gene_names), size = 3)\n",
        "\n",
        "ggsave(filename = \"./volcano plot.tiff\", plot = my_plot_1, width = 8, height = 6, dpi = 300,\n",
        "       device = tiff,\n",
        "       bg = \"white\")\n",
        "dev.off()\n",
        "\n",
        "ggsave(filename = \"./volcano plot.pdf\", plot = my_plot_1, width = 8, height = 6, dpi = 300,\n",
        "       device = pdf(),\n",
        "       bg = \"white\")\n",
        "dev.off()\n",
        "\n",
        "\n",
        "# Heat map\n",
        "top_genes <- rownames(head(ttest_res[order(ttest_res$adj.P.Val), ], 100)) # 상위 100개 유전자 선택\n",
        "top_genes_data <- y[top_genes, ] # 상위 유전자에 대한 발현 데이터 선택\n",
        "\n",
        "my_palette <- colorRampPalette(c(\"navy\", \"white\", \"red\"))(50)\n",
        "col_annotation <- design\n",
        "rownames(col_annotation) <- col_annotation[,1]\n",
        "col_annotation[ , 1] <- NULL\n",
        "my_plot_2 <- pheatmap(top_genes_data, cluster_rows = TRUE, cluster_cols = TRUE, annotation_col = col_annotation,\n",
        "                      scale = \"row\", color = my_palette, show_rownames = FALSE, show_colnames = FALSE, main = \"Heat Map\")\n",
        "\n",
        "ggsave(filename = \"./heatmap.tiff\", plot = my_plot_2, width = 8, height = 6, dpi = 300,\n",
        "       device = tiff)\n",
        "dev.off()\n",
        "\n",
        "ggsave(filename = \"./heatmap.pdf\", plot = my_plot_2, width = 8, height = 6, dpi = 300,\n",
        "       device = pdf())\n",
        "dev.off()\n",
        "\n",
        "\n",
        "# PCA\n",
        "pca_result <- prcomp(t(log2(y + 1)))\n",
        "pca_df <- data.frame(PC1 = pca_result$x[, 1], PC2 = pca_result$x[, 2], Group = group)\n",
        "my_plot_3 <- ggplot(pca_df, aes(x = PC1, y = PC2, color = Group)) + geom_point()\n",
        "\n",
        "ggsave(filename = \"./PCA.tiff\", plot = my_plot_3, width = 8, height = 6, dpi = 300,\n",
        "       device = tiff,\n",
        "       type = \"cairo\")\n",
        "dev.off()\n",
        "\n",
        "ggsave(filename = \"./PCA.pdf\", plot = my_plot_3, width = 8, height = 6, dpi = 300,\n",
        "       device = pdf(),\n",
        "       bg = \"white\")\n",
        "dev.off()\n",
        "\n",
        "# 6. Enrichment Analysis\n",
        "# get the significant up-regulated/down-regulated genes\n",
        "up <- ttest_res[ttest_res$adj.P.Val < 0.05 & ttest_res$logFC > 1,]\n",
        "down <- ttest_res[ttest_res$adj.P.Val < 0.05 & ttest_res$logFC < -1,]\n",
        "\n",
        "# order genes by adjusted-pvalue\n",
        "up_ordered = up[order(up$adj.P.Val, decreasing = FALSE),]\n",
        "down_ordered = down[order(down$adj.P.Val, decreasing = FALSE),]\n",
        "\n",
        "# ordered enrichment analysis\n",
        "gp_up_ordered = gost(list(\"up-regulated\" = row.names(up_ordered)), organism = \"hsapiens\", source = c(\"GO:BP\", \"GO:MF\", \"GO:CC\", \"KEGG\"))\n",
        "gp_down_ordered = gost(list(\"down-regulated\" = row.names(down_ordered)), organism = \"hsapiens\", source = c(\"GO:BP\", \"GO:MF\", \"GO:CC\", \"KEGG\"))\n",
        "\n",
        "# Enrichment plot\n",
        "top_BP_terms_up <- head(gp_up_ordered$result[gp_up_ordered$result$source == \"GO:BP\", \"term_id\"], 10)\n",
        "top_MF_terms_up <- head(gp_up_ordered$result[gp_up_ordered$result$source == \"GO:MF\", \"term_id\"], 10)\n",
        "top_CC_terms_up <- head(gp_up_ordered$result[gp_up_ordered$result$source == \"GO:CC\", \"term_id\"], 10)\n",
        "top_KEGG_terms_up <- head(gp_up_ordered$result[gp_up_ordered$result$source == \"KEGG\", \"term_id\"], 10)\n",
        "top_terms_up <- c(top_BP_terms_up, top_MF_terms_up, top_CC_terms_up, top_KEGG_terms_up)\n",
        "\n",
        "top_BP_terms_down <- head(gp_down_ordered$result[gp_down_ordered$result$source == \"GO:BP\", \"term_id\"], 10)\n",
        "top_MF_terms_down <- head(gp_down_ordered$result[gp_down_ordered$result$source == \"GO:MF\", \"term_id\"], 10)\n",
        "top_CC_terms_down <- head(gp_down_ordered$result[gp_down_ordered$result$source == \"GO:CC\", \"term_id\"], 10)\n",
        "top_KEGG_terms_down <- head(gp_down_ordered$result[gp_down_ordered$result$source == \"KEGG\", \"term_id\"], 10)\n",
        "top_terms_down <- c(top_BP_terms_down, top_MF_terms_down, top_CC_terms_down, top_KEGG_terms_down)\n",
        "\n",
        "p1 <- gostplot(gp_up_ordered, interactive = FALSE)\n",
        "publish_gostplot(p1, highlight_terms = top_terms_up, width = 10, height = 20, filename = \"upregulated_enrichment.pdf\")\n",
        "\n",
        "p2 <- gostplot(gp_down_ordered, interactive = FALSE)\n",
        "publish_gostplot(p2, highlight_terms = top_terms_down, width = 10, height = 20, filename = \"downregulated_enrichment.pdf\")\n",
        "\n",
        "# multi_gp_link <- gost(list(\"up-regulated\" = row.names(up_ordered), \"down-regulated\" = row.names(down_ordered)),\n",
        "#                      organism='hsapiens', source = c(\"GO:BP\", \"GO:MF\", \"GO:CC\", \"KEGG\"), as_short_link = TRUE)\n",
        "\n",
        "\n",
        "# 7. 결과를 csv 파일로 저장\n",
        "write.csv(ttest_res, \"ttest_DEG_results.csv\")"
      ],
      "metadata": {
        "id": "jGMvJ2N_3NBJ"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}